rules/directories/providers.bzl

```diff
@@ -0,0 +1,33 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Skylib module containing providers for directories."""
+
+visibility("public")
+
+DirectoryInfo = provider(
```

Comment:
This provider looks surprisingly heavyweight. Presumably, someone reaching for this kind of rule would likely want to just do something like:
```
ctx.actions.run(
    executable = [ctx.executable.sphinx],
    inputs = ctx.docs_root["DefaultInfo"].files.to_list(),
    outputs = [out_dir],
    arguments = [
        ctx.docs_root["DirectoryInfo"].directory.
        out_dir.path,
    ]
)
```

With the current provider's structure, how to achieve that goal is non-obvious to me. I understand that a lot of this information is handy (and in some cases critical to the current implementation), but it seems to get in the way of usability.

Reply:
FYI, I created [rules_directory](https://github.com/matts1/rules_directory/tree/main) to do this instead. I'll close this PR.

And I agree with your concerns, when rewriting for rules_directory I simplified it a lot. The following should now be possible:

```
directory(
    name = "docs_root",
    srcs = glob(["**/*"], exclude=["BUILD.bazel"])
)

foo_rule(
    ...,
    docs_root = ":docs_root",
)
```

Then in the foo_rule implementation:
```
ctx.actions.run(
    executable = [ctx.executable._sphinx],
    inputs = ctx.docs_root[DefaultInfo].files.to_list(),
    outputs = [out_dir],
    arguments = [
        # This is preferred. It auto-detects if it's a source directory or a generated directory and picks appropriately.
        directory_path(ctx.docs_root[DirectoryInfo]),
        # If it can't tell which it is, one of these may be required
        ctx.docs_root[DirectoryInfo].source_path,
        ctx.docs_root[DirectoryInfo].generated_path,
        out_dir.path,
    ]
)
```